### ISSUE TITLE
Request for a casual AttentionBias mask with qkv padding

### ISSUE THREAD

**Issue by peterjc123** [facebookresearch/xformers#782]
# 🚀 Feature

We need one kind of AttentionBias like BlockDiagonalCausalMask, but with some optional padding.

## Motivation

When training LLM, training data may be packed. It may look like

```
A A A A B B B .... D D PAD PAD PAD
```

We need the attention mask to ignore the padding part of the input data.

**Comment by danthe3rd** [comment]
Hi @peterjc123,

At the moment, we only have [`BlockDiagonalCausalWithOffsetPaddedKeysMask`](https://github.com/facebookresearch/xformers/blob/55a4798d99c350eca9d225a14003f226775e5f76/xformers/ops/fmha/attn_bias.py#L528) that supports padding, with the following limitations:
(1) Only supported for inference (eg no backward pass)
(2) Only the key/value are padded - the queries are packed
This is mostly meant for inference where you have a key/value cache of fixed size.

May I ask what your use-case is, and why `BlockDiagonalCausalMask` does not work for you? Are both queries/keys/values padded?
If the padding is only at the end of the sequences, I'm wondering if something like this could work for you:
```
# Q/K/V are:
# A A B B B C PAD PAD PAD ...
# NOTE: Ignore the padding
# NOTE: You will get garbage output in the padded section, but I guess it's fine?
bias = BlockDiagonalMask.from_seqlens([2, 3, 1]).make_causal()
```

**Comment by peterjc123** [comment]
> May I ask what your use-case is, and why `BlockDiagonalCausalMask` does not work for you?
I want to use XFormers to speed up LLM trainings, especially the self-attention part. For example, given the texts "ab" "gh" and 6 as max sequence length, so that the packed input data will be "abgh<PAD><PAD>". And the attention mask will be something like below.
```
1 0 0 0 0 0
1 1 0 0 0 0
0 0 1 0 0 0
0 0 1 1 0 0
0 0 0 0 0 0
0 0 0 0 0 0
```
But using `BlockDiagonalCausalMask`, you may get something like below.
```
```
1 0 0 0 0 0
1 1 0 0 0 0
0 0 1 0 0 0
0 0 1 1 0 0
0 0 0 0 1 0
0 0 0 0 0 1
```



> Are both queries/keys/values padded?

Yes, for LLM training with packed data, the queries, keys and values are all padded.



> If the padding is only at the end of the sequences, maybe you can just slice it to remove the last part before you pass it to the attention?

Yes, I can try this but I guess this will lead to degraded performance. Since pad operations are quite expensive.

**Comment by danthe3rd** [comment]
Padding will involve a copy of the vector. It shouldn't be that expensive if your model is large enough.
You might as well want to try this trick - should work in theory, but maybe we are guarding against this somewhere:
```
# Q/K/V are:
# A A B B B C PAD PAD PAD ...
bias = BlockDiagonalMask.from_seqlens([2, 3, 1]).make_causal()

# You will get garbage output/gradients in the padded area, hopefully it does not matter?
```

**Comment by peterjc123** [comment]
> `bias = BlockDiagonalMask.from_seqlens([2, 3, 1]).make_causal()`

I guess this won't work well if the input data is batched as well. For batched data,  padding & slicing operations are also difficult to apply.

**Comment by peterjc123** [comment]
Using `BlockDiagonalCausalMask.from_seqlens(torch.unique_consecutive(attention_mask, return_counts=True)[1].cpu().tolist())` is a current workaround. The training process can be carried out and the losses are similar but I still believe it's not 100% equivalent.
